File: api/schemas/task.py
"""
任务相关数据模型
"""
from enum import Enum
from typing import Optional, Dict, Any, List
from datetime import datetime, timedelta
from pydantic import BaseModel, Field
import uuid


class TaskStatus(str, Enum):
    """任务状态"""
    PENDING = "pending"          # 等待执行
    RUNNING = "running"          # 执行中
    COMPLETED = "completed"      # 执行完成
    FAILED = "failed"            # 执行失败
    CANCELLED = "cancelled"      # 已取消


class TaskPriority(int, Enum):
    """任务优先级"""
    LOW = 1
    NORMAL = 5
    HIGH = 8
    URGENT = 10


class TaskConfig(BaseModel):
    """任务配置"""
    # 基本配置
    platform: str                                       # xhs/dy/bili/wb/wechat/ks/tieba/zhihu
    crawler_type: str = "search"                        # search/creator/detail/creator_vip/album
    
    # 搜索配置
    keywords: Optional[List[str]] = None                # 搜索关键词列表
    
    # 创作者配置
    creator_ids: Optional[List[str]] = None             # 创作者 ID 或 URL 列表
    
    # 详情配置
    note_urls: Optional[List[str]] = None               # 笔记/视频 URL 列表
    
    # 爬取数量配置
    max_notes: int = 100                                # 最大爬取笔记数
    enable_comments: bool = False                       # 是否爬取评论
    max_comments_per_note: int = 20                     # 每个笔记最大评论数
    enable_media: bool = False                          # 是否下载图片/视频
    
    # 执行配置
    concurrency: int = 3                                # 并发数
    crawl_interval: float = 1.0                         # 请求间隔（秒）
    
    # 登录配置
    login_type: Optional[str] = "cookie"                # cookie/qrcode
    cookies: Optional[str] = None                       # Cookie 字符串
    
    # 存储配置
    save_option: str = "json"                           # json/csv/excel/db/sqlite
    
    # 可扩展字段
    extra: Dict[str, Any] = Field(default_factory=dict)


class TaskProgress(BaseModel):
    """任务进度"""
    current: int = 0
    total: int = 0
    percentage: int = 0
    items_crawled: int = 0
    comments_crawled: int = 0


class TaskResult(BaseModel):
    """任务结果"""
    success: bool = False
    error_message: Optional[str] = None
    output_path: Optional[str] = None
    statistics: Dict[str, Any] = Field(default_factory=dict)


class TaskMetadata(BaseModel):
    """任务元数据"""
    sign_server_used: bool = False
    execution_node: Optional[str] = None       # 分布式时的执行节点
    resource_usage: Dict[str, Any] = Field(default_factory=dict)


class Task(BaseModel):
    """任务实体"""
    task_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    session_id: str
    task_name: Optional[str] = None
    
    # 状态
    status: TaskStatus = TaskStatus.PENDING
    priority: int = TaskPriority.NORMAL
    retry_count: int = 0
    max_retries: int = 3
    idempotency_key: Optional[str] = None      # 幂等键
    cancel_requested: bool = False             # 取消标记
    
    # 时间
    created_at: datetime = Field(default_factory=datetime.utcnow)
    scheduled_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    last_heartbeat_at: Optional[datetime] = None
    
    # 详情
    config: TaskConfig
    progress: TaskProgress = Field(default_factory=TaskProgress)
    result: TaskResult = Field(default_factory=TaskResult)
    metadata: TaskMetadata = Field(default_factory=TaskMetadata)
    
    # 版本号（用于乐观锁）
    version: int = 0
    
    class Config:
        use_enum_values = True


class TaskLease(BaseModel):
    """任务租约（用于 reserve/ack/nack）"""
    lease_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    task_id: str
    worker_id: str
    acquired_at: datetime = Field(default_factory=datetime.utcnow)
    expires_at: datetime
    
    def is_expired(self) -> bool:
        return datetime.utcnow() > self.expires_at
    
    @classmethod
    def create(cls, task_id: str, worker_id: str, lease_seconds: int = 300) -> "TaskLease":
        """创建租约"""
        now = datetime.utcnow()
        return cls(
            task_id=task_id,
            worker_id=worker_id,
            acquired_at=now,
            expires_at=now + timedelta(seconds=lease_seconds)
        )


# ========== 请求/响应模型 ==========

class TaskCreateRequest(BaseModel):
    """创建任务请求"""
    task_name: Optional[str] = None
    config: TaskConfig
    priority: int = TaskPriority.NORMAL
    scheduled_at: Optional[datetime] = None
    idempotency_key: Optional[str] = None


class TaskListRequest(BaseModel):
    """任务列表查询"""
    status: Optional[TaskStatus] = None
    platform: Optional[str] = None
    page: int = 1
    page_size: int = 20


class TaskListResponse(BaseModel):
    """任务列表响应"""
    tasks: List[Task]
    total: int
    page: int
    page_size: int


class TaskStatsResponse(BaseModel):
    """任务统计响应"""
    pending: int = 0
    running: int = 0
    completed: int = 0
    failed: int = 0
    cancelled: int = 0
    total: int = 0

File: webui-src/src/types/task.ts
/**
 * 任务相关类型定义
 */

// 任务状态
export type TaskStatus = 
  | 'pending' 
  | 'running' 
  | 'completed' 
  | 'failed' 
  | 'cancelled';

// 平台
export type Platform = 'xhs' | 'dy' | 'bili' | 'wb' | 'wechat' | 'ks' | 'tieba' | 'zhihu';

// 爬取类型
export type CrawlerType = 'search' | 'creator' | 'detail' | 'creator_vip' | 'album';

// 日志级别
export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

// 任务配置
export interface TaskConfig {
  // 基本配置
  platform: Platform;
  crawler_type: CrawlerType;
  
  // 搜索配置
  keywords?: string[];                         // 搜索关键词列表
  
  // 创作者配置
  creator_ids?: string[];                      // 创作者 ID 或 URL 列表
  
  // 详情配置
  note_urls?: string[];                        // 笔记/视频 URL 列表
  
  // 爬取数量配置
  max_notes: number;                           // 最大爬取笔记数
  enable_comments: boolean;                    // 是否爬取评论
  max_comments_per_note?: number;              // 每个笔记最大评论数
  enable_media?: boolean;                      // 是否下载图片/视频
  
  // 执行配置
  concurrency?: number;                        // 并发数
  crawl_interval?: number;                     // 请求间隔（秒）
  
  // 登录配置
  login_type?: 'cookie' | 'qrcode';            // 登录类型
  cookies?: string;                            // Cookie 字符串
  
  // 存储配置
  save_option: 'csv' | 'json' | 'excel' | 'db' | 'sqlite';
  
  // 可扩展字段
  extra?: Record<string, unknown>;
}

// 任务进度
export interface TaskProgress {
  current: number;
  total: number;
  percentage: number;
  items_crawled: number;
  comments_crawled: number;
}

// 任务结果
export interface TaskResult {
  success: boolean;
  error_message?: string;
  output_path?: string;
  statistics?: Record<string, unknown>;
}

// 任务元数据
export interface TaskMetadata {
  sign_server_used: boolean;
  execution_node?: string;
  resource_usage?: Record<string, unknown>;
}

// 任务实体
export interface Task {
  task_id: string;
  session_id: string;
  task_name?: string;
  status: TaskStatus;
  priority: number;
  retry_count: number;
  max_retries: number;
  idempotency_key?: string;
  cancel_requested: boolean;
  created_at: string;
  scheduled_at?: string;
  started_at?: string;
  finished_at?: string;
  last_heartbeat_at?: string;
  config: TaskConfig;
  progress: TaskProgress;
  result: TaskResult;
  metadata: TaskMetadata;
  version: number;
}

// 创建任务请求
export interface TaskCreateRequest {
  task_name?: string;
  config: TaskConfig;
  priority?: number;
  scheduled_at?: string;
  idempotency_key?: string;
}

// 任务列表查询
export interface TaskListRequest {
  status?: TaskStatus;
  platform?: Platform;
  page?: number;
  page_size?: number;
}

// 任务列表响应
export interface TaskListResponse {
  tasks: Task[];
  total: number;
  page: number;
  page_size: number;
}

// 任务统计
export interface TaskStats {
  pending: number;
  running: number;
  completed: number;
  failed: number;
  cancelled: number;
  total: number;
}

// 日志条目
export interface LogEntry {
  log_id: string;
  task_id: string;
  timestamp: string;
  level: LogLevel;
  message: string;
  extra?: Record<string, unknown>;
}

// 任务事件
export interface TaskEvent {
  event_id: string;
  event_type: string;
  task_id: string;
  session_id: string;
  timestamp: string;
  payload: Record<string, unknown>;
}

// 会话配额
export interface SessionQuota {
  max_concurrent_tasks: number;
  max_daily_tasks: number;
  used_daily_tasks: number;
  quota_reset_at: string;
}

// 会话响应
export interface SessionResponse {
  session_id: string;
  user_id?: string;
  created_at: string;
  expires_at: string;
  quota: SessionQuota;
}

// 平台信息
export interface PlatformInfo {
  value: Platform;
  label: string;
  icon?: string;
}

// 爬取类型信息
export interface CrawlerTypeInfo {
  value: CrawlerType;
  label: string;
}

// 优先级
export const TaskPriority = {
  LOW: 1,
  NORMAL: 5,
  HIGH: 8,
  URGENT: 10,
} as const;

// 状态显示配置
export const TaskStatusConfig: Record<TaskStatus, { label: string; color: string; bgColor: string }> = {
  pending: { label: '等待中', color: 'text-gray-700', bgColor: 'bg-gray-100' },
  running: { label: '运行中', color: 'text-blue-700', bgColor: 'bg-blue-100' },
  completed: { label: '已完成', color: 'text-green-700', bgColor: 'bg-green-100' },
  failed: { label: '已失败', color: 'text-red-700', bgColor: 'bg-red-100' },
  cancelled: { label: '已取消', color: 'text-yellow-700', bgColor: 'bg-yellow-100' },
};

// 平台显示配置
export const PlatformConfig: Record<Platform, PlatformInfo> = {
  xhs: { value: 'xhs', label: '小红书', icon: 'book-open' },
  dy: { value: 'dy', label: '抖音', icon: 'music' },
  bili: { value: 'bili', label: 'B站', icon: 'tv' },
  wb: { value: 'wb', label: '微博', icon: 'message-circle' },
  wechat: { value: 'wechat', label: '微信公众号', icon: 'message-square' },
  ks: { value: 'ks', label: '快手', icon: 'video' },
  tieba: { value: 'tieba', label: '百度贴吧', icon: 'messages-square' },
  zhihu: { value: 'zhihu', label: '知乎', icon: 'help-circle' },
};

// 爬取类型配置
export const CrawlerTypeConfig: Record<CrawlerType, CrawlerTypeInfo> = {
  search: { value: 'search', label: '关键词搜索' },
  creator: { value: 'creator', label: '创作者主页' },
  detail: { value: 'detail', label: '帖子详情' },
  creator_vip: { value: 'creator_vip', label: 'VIP内容（微博）' },
  album: { value: 'album', label: '合集模式（微信）' },
};

